### PR TITLE
Initial implementation of terraform-aws-vpc

### PR DESCRIPTION
**Pull request by rafikurnia**
This module was tested on dev account. Please review. Thank you..

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+    principals = {
+      type        = "Service"
+      identifiers = ["vpc-flow-logs.amazonaws.com"]
+    }
+  }
+}
+
+# Provides an IAM role for VPC Flow Logs.
+resource "aws_iam_role" "flow_logs" {
+  name        = "ServiceRoleForVPCFlowLogs_${var.vpc_name}"
```
Comment (sshvetsov) [inline main.tf]: I advise to add RANDOM_ID at the end

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+  description = "Service Role for VPC Flow Logs - ${var.vpc_name} VPC"
+  path        = "/service-role/vpc-flow-logs.amazonaws.com/"
+
+  assume_role_policy = "${data.aws_iam_policy_document.flow_logs_trust_policy.json}"
+}
+
+# Provides a CloudWatch Log Group resource for VPC Flow Logs to store the logs.
+resource "aws_cloudwatch_log_group" "flow_logs" {
+  name              = "/aws/vpc-flow-logs/${var.vpc_name}"
```
Comment (sshvetsov) [inline main.tf]: I advise to add RANDOM_ID to the end of the name

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+resource "aws_iam_role" "flow_logs" {
+  name        = "ServiceRoleForVPCFlowLogs_${var.vpc_name}"
+  description = "Service Role for VPC Flow Logs - ${var.vpc_name} VPC"
+  path        = "/service-role/vpc-flow-logs.amazonaws.com/"
+
+  assume_role_policy = "${data.aws_iam_policy_document.flow_logs_trust_policy.json}"
+}
+
+# Provides a CloudWatch Log Group resource for VPC Flow Logs to store the logs.
+resource "aws_cloudwatch_log_group" "flow_logs" {
```
Comment (sshvetsov) [inline main.tf]: Let's also add the Log group to outputs

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  name              = "/aws/vpc-flow-logs/${var.vpc_name}"
+  retention_in_days = 0
```
Comment (sshvetsov) [inline main.tf]: This should probably be configurable via variables

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+    Environment   = "${var.environment}"
+    Description   = "${var.environment} VPC for ${var.product_domain} product domain"
+    ManagedBy     = "Terraform"
+  }
+}
+
+# Provides VPC public subnet resources (DMZ).
+# One for each AZ.
+resource "aws_subnet" "public" {
+  count = "${length(var.subnet_availability_zones)}"
```
[the remark this answers is not in the document]
Comment (rafikurnia) [reply main.tf]: I wrote the note about this at README.md:
***This module supports only up to 4 AZs.***

Another good thing is Terraform (the `cidrsubnet` interpolation function) will also prevent the stack from being created at `terraform plan` stage. I tried to input 5 AZs:
```bash
Error: module.dev.aws_subnet.app: 1 error(s) occurred:

* module.dev.aws_subnet.app[4]: cidrsubnet: prefix extension of 3 does not accommodate a subnet numbered 8 in:

${cidrsubnet(var.vpc_cidr_block, 3, count.index + 4)}
```

Do we still need to limit all counts to 4? It is possible using `min` interpolation function:
```hcl
count = "${min(4, length(var.subnet_availability_zones))}"
```

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+    MultiTier     = "${var.vpc_multi_tier ? "true" : "false"}"
+    ProductDomain = "${var.product_domain}"
+    Environment   = "${var.environment}"
+    Description   = "${var.environment} VPC for ${var.product_domain} product domain"
+    ManagedBy     = "Terraform"
+  }
+}
+
+# Provides VPC public subnet resources (DMZ).
+# One for each AZ.
```
[the remark this answers is not in the document]
Comment (rafikurnia) [reply main.tf]: I see.. I manually tested it to switch between multi-tier and single-tier by changing  `vpc_multi_tier` variable though, and it works. But I got your point. Should we remove this feature and make this module for 3-tier VPC only? We can always create another module for single-tier-VPC-only later..

What do you think about this @andysaputra ?

[the remark this answers is not in the document]
Comment (rafikurnia) [reply main.tf]: I need input from @traveloka/site-infra team as well :bowing_man:

File: main.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (arief-halim) [reply main.tf]: Using interpolation syntax **min** will hide the actual issue when user input 5 AZ (there is no exception thrown)

[the remark this answers is not in the document]
Comment (sshvetsov) [reply main.tf]: I prefer explicit error to silently ignoring invalid value. POLA.

File: main.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (sshvetsov) [reply main.tf]: I think simpler is better. Lets make it a simple 3-tier module if time allows.

**Pull request merged**